### PR TITLE
docs: Use and define languages

### PR DESCRIPTION
**Pull request by nzakas**
#### Prerequisites checklist

- [x] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/HEAD/CONTRIBUTING.md).

#### What is the purpose of this pull request? (put an "X" next to an item)

[x] Documentation update
[ ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-change-proposal.md))
[ ] Add autofix to a rule
[ ] Add a CLI option
[ ] Add something to the core
[ ] Other, please explain:

#### What changes did you make? (Give an overview)

Adds some basic documentation around:

1. How to use languages from plugins
2. How to create languages

#### Is there anything you'd like reviewers to focus on?

Is this enough to get people started? We could probably write a small book on how to create languages, but I'm just trying to get the basics written down for now and hoping that the examples of the JSON plugin, and soon the Markdown plugin, will be enough to get people going.

### CODE REVIEW

File: docs/src/extend/languages.md [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+* `nodeTypeKey` - the name of the property that indicates the node type (usually `"type"` or `"kind"`).
+* `validateLanguageOptions(languageOptions)` - validates language options for the language. This method is expected to throw a validation error when an expected language option doesn't have the correct type or value. Unexpected language options should be silently ignored and no error should be thrown. This method is required eve if the language doesn't specify any options.
+* `parse(file, context)` - parses the given file into an AST or CST, and can also include additional values meant for use in rules. Called internally by ESLint.
+* `createSourceCode(file, parseResult, context)` - creates a `SourceCode` object. Call internally by ESLint after `parse()`, and the second argument is the exact return value from `parse()`.
+
+The following optional members allow you to customize how ESLint interacts with the object:
+
+* `visitorKeys` - visitor keys that are specific the AST or CST. This is used to optimize traversal of the AST or CST inside of ESLint. While not required, it is strongly recommended, especially for AST or CST formats that deviate significantly from ESTree format.
```
Comment (amareshsm) [inline docs/src/extend/languages.md:83]:
```suggestion
* `visitorKeys` - visitor keys that are specific to the AST or CST. This is used to optimize traversal of the AST or CST inside of ESLint. While not required, it is strongly recommended, especially for AST or CST formats that deviate significantly from ESTree format.
```

File: docs/src/extend/languages.md [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+
+A basic `Language` object must implement the following:
+
+* `fileType` - should be `"text"` (in the future, we will also support `"binary"`)
+* `lineStart` - either 0 or 1 to indicate how the AST represents the first line in the file. ESLint uses this to correctly display error locations.
+* `columnStart` - either 0 or 1 to indicate how the AST represents the first column in the file. ESLint uses this to correctly display error locations.
+* `nodeTypeKey` - the name of the property that indicates the node type (usually `"type"` or `"kind"`).
+* `validateLanguageOptions(languageOptions)` - validates language options for the language. This method is expected to throw a validation error when an expected language option doesn't have the correct type or value. Unexpected language options should be silently ignored and no error should be thrown. This method is required eve if the language doesn't specify any options.
```
Comment (amareshsm) [inline docs/src/extend/languages.md:77]:
```suggestion
* `validateLanguageOptions(languageOptions)` - validates language options for the language. This method is expected to throw a validation error when an expected language option doesn't have the correct type or value. Unexpected language options should be silently ignored and no error should be thrown. This method is required even if the language doesn't specify any options.
```

**Comment by nzakas** [comment]
@amareshsm just a reminder -- when you start reviewing a PR, it should be moved to "Implementing" on the Triage board.

File: docs/src/extend/languages.md [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+
+ESLint holds information about source code in a `SourceCode` object. This object is the API used both by ESLint internally and by rules written to work on the code (via `context.sourceCode`). The `SourceCode` object must implement the `TextSourceCode` interface as defined in the [`@eslint/core`](https://npmjs.com/package/@eslint/core) package.
+
+A basic `SourceCode` object must implement the following:
+
+* `ast` - a property containing the AST or CST for the source code.
+* `text` - the text of the source code.
+* `getLoc(nodeOrToken)` - a method that returns the location of a given node or token. This must match the `loc` structure that ESTree uses.
+* `getRange(nodeOrToken)` - a method that returns the range of a given node or token. This must return an array where the first item is the start index and the second is the end index.
+* `traverse()` - a method that returns an iterator for traversing the AST or CST. The iterator must return objects that implement either `VisitTraversalStep` or `CallTraversalStep` from `@eslint/core`.
```
Comment (mdjermanovic) [inline docs/src/extend/languages.md:47]:
```suggestion
* `traverse()` - a method that returns an iterable for traversing the AST or CST. The iterator must return objects that implement either `VisitTraversalStep` or `CallTraversalStep` from `@eslint/core`.
```

`traverse()` should return an iterable object (which we then use with for-of), not an iterator.

File: docs/src/extend/languages.md [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+
+### The `Language` Object
+
+The `Language` object contains all of the information about the programming language as well as methods for interacting with code written in that language. ESLint uses this object to determine how to deal with a particular file. The `Language` object must implement the `Language` interface as defined in the [`@eslint/core`](https://npmjs.com/package/@eslint/core) package.
+
+A basic `Language` object must implement the following:
+
+* `fileType` - should be `"text"` (in the future, we will also support `"binary"`)
+* `lineStart` - either 0 or 1 to indicate how the AST represents the first line in the file. ESLint uses this to correctly display error locations.
+* `columnStart` - either 0 or 1 to indicate how the AST represents the first column in the file. ESLint uses this to correctly display error locations.
```
Comment (mdjermanovic) [inline docs/src/extend/languages.md:75]:
```suggestion
* `columnStart` - either 0 or 1 to indicate how the AST represents the first column in each line. ESLint uses this to correctly display error locations.
```


I think it's clearer this way.

File: docs/src/extend/languages.md [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+const plugin = {
+
+    // preferred location of name and version
+    meta: {
+        name: "eslint-plugin-example",
+        version: "1.2.3"
+    },
+    languages: {
+        my: myLanguage
+    }
```
Comment (mdjermanovic) [inline docs/src/extend/languages.md:104]:
```suggestion
    },
```

**Review by mdjermanovic — APPROVED** [review]
LGTM, thanks!

**Pull request merged**